temporalio/worker/workflow_instance.py

```diff
@@ -0,0 +1,1396 @@
+                        command.respond_to_query.failed,
+                    )
+                    self._pending_commands.append(command)
+                except Exception as inner_err:
+                    raise ValueError(
+                        "Failed converting application error"
+                    ) from inner_err
+
+        # Just find the arg types for now. The interceptor will be responsible
+        # for checking whether the query definition actually exists.
+        query_defn = self._queries.get(job.query_type)
+        arg_types, _ = (
+            self._type_lookup.get_type_hints(query_defn.fn)
+            if query_defn
+            else (None, None)
+        )
+        args = self._convert_payloads(job.arguments, arg_types)
+
+        # Schedule it
+        self.create_task(
+            run_query(
+                HandleQueryInput(
+                    id=job.query_id,
+                    name=job.query_type,
+                    args=args,
+                )
+            )
+        )
+
+    def _apply_resolve_activity(
+        self, job: temporalio.bridge.proto.workflow_activation.ResolveActivity
+    ) -> None:
+        handle = self._pending_activities.get(job.seq)
+        if not handle:
+            raise RuntimeError(f"Failed finding activity handle for sequence {job.seq}")
+        if job.result.HasField("completed"):
+            self._pending_activities.pop(job.seq)
+            ret: Optional[Any] = None
+            if job.result.completed.HasField("result"):
+                ret_types = [handle._input.ret_type] if handle._input.ret_type else None
+                ret_vals = self._convert_payloads(
+                    [job.result.completed.result],
+                    ret_types,
+                )
+                ret = ret_vals[0]
+            handle._resolve_success(ret)
+        elif job.result.HasField("failed"):
+            self._pending_activities.pop(job.seq)
+            handle._resolve_failure(
+                temporalio.exceptions.failure_to_error(
+                    job.result.failed.failure, self._payload_converter
+                )
+            )
+        elif job.result.HasField("cancelled"):
+            self._pending_activities.pop(job.seq)
+            handle._resolve_failure(
+                temporalio.exceptions.failure_to_error(
+                    job.result.cancelled.failure, self._payload_converter
+                )
+            )
+        elif job.result.HasField("backoff"):
+            # Async call to wait for the timer then execute the activity again
+            async def backoff_then_reschedule() -> None:
+                # Don't pop, we want this to remain as pending
+                handle = self._pending_activities[job.seq]
+                try:
+                    await asyncio.sleep(
+                        job.result.backoff.backoff_duration.ToTimedelta().total_seconds()
+                    )
+                except Exception as err:
```

Comment:
When would `sleep` throw?

Reply:
Looks like an unexpected error

Reply:
Cancel I believe. I need to test LA backoff cancel anyways.

Reply:
AFAIR CancelledError doesn't subclass Exception in recent Python versions.
Looks like this is WIP so not pressing this further.

Reply:
> AFAIR CancelledError doesn't subclass Exception in recent Python versions.

True, only `BaseException`, ref https://docs.python.org/3/library/asyncio-exceptions.html#asyncio.CancelledError:

> Changed in version 3.8: [CancelledError](https://docs.python.org/3/library/asyncio-exceptions.html#asyncio.CancelledError) is now a subclass of [BaseException](https://docs.python.org/3/library/exceptions.html#BaseException).

I need to perform more tests and maybe change the exception type here. I do need to support cancelling a LA backoff timer that reflects as the LA task failing, but I think I may need to rework the LA to have this backoff happen inside the task instead of separate like it does here (e.g. with an event).

Reply:
I have completed reworked LA backoff. I now resolve the LA future with a special exception telling it to backoff and try again. This also means the sleep is inside the LA task so if the task is cancelled during backoff, the timer is properly cancelled. I still haven't added tests for this though.